module/zfs/arc.c

```diff
@@ -5815,7 +5818,9 @@ arc_read_done(zio_t *zio)
 		    acb->acb_compressed, acb->acb_noauth, B_TRUE,
 		    &acb->acb_buf);
 		if (error != 0) {
-			arc_buf_destroy(acb->acb_buf, acb->acb_private);
```

Comment:
Unfortunately that doesn't work (unless I add some more exception code). `arc_buf_destroy_impl()` calls `arc_cksum_verify()` which will panic debug builds if the IO error bit isn't set (since the arc buffer is empty and therefore does not match the header checksum).